Implement plugin enable / disable

mem opened the pull request:
Implement all the plugin enable / disable functionality.

- Fixes #2322 

Signed-off-by: Marcelo E. Magallon <marcelo@sylabs.io>

Commit:
Implement all the plugin enable / disable functionality

- Fixes #2322

Signed-off-by: Marcelo E. Magallon <marcelo@sylabs.io>

bauerm97 commented on cmd/internal/cli/plugin_enable_linux.go:
```diff
@@ -0,0 +1,34 @@
+		if err != nil {
+			if os.IsNotExist(err) {
+				sylog.Fatalf("Failed to enable plugin %q: plugin not found.", args[0])
+			} else {
```
`} else {` not necessary here. Might be simpler as:

```
if os.IsNotExist(err) {
    sylog.Fatalf([...])
}
sylog.Fatalf([...])
```

mem replied:
I was trying to make it non-ambiguous that it's handling "all the other errors" without forcing the reader to keep in mind that "sylog.Fatalf" interrupts the program, but I can change this. Originally, I had a switch here, but since it's only one special case, it looks weird.

mem replied:
Changed it to what @bauerm97 is asking for.

bauerm97 commented on cmd/internal/cli/plugin_disable_linux.go:
```diff
@@ -0,0 +1,34 @@
+		if err != nil {
+			if os.IsNotExist(err) {
+				sylog.Fatalf("Failed to disable plugin %q: plugin not found.", args[0])
+			} else {
```
Same comment

bauerm97 commented on internal/pkg/plugin/binary.go:
```diff
@@ -149,6 +182,17 @@ func loadMetaByName(name, plugindir string) (*Meta, error) {
+func loadMetaByFilename(filename string) (*Meta, error) {
+	fh, err := os.Open(filename)
+	if err != nil {
+		sylog.Debugf("Error opening meta file %q: %s\n", filename, err)
```
`sylog.Errorf`

mem replied:
I don't actually want that. The debug entry is there just in case it's needed (you know, for debugging :-) but for the user I would like to have a different error. That's why I'm checking whether it's a not exist error elsewhere.

bauerm97 commented on internal/pkg/plugin/binary.go:
```diff
@@ -120,23 +120,56 @@ func Uninstall(name, libexecdir string) error {
-func loadMetaByName(name, plugindir string) (*Meta, error) {
-	fh, err := os.Open(metaPath(plugindir, name))
+// Enable enables the plugin named "name" found under "libexecdir"
+func Enable(name, libexecdir string) error {
```
It does seem like these functions here are more like "application code" rather than the underlying backing package. I think that the `internal/app/singularity/remote_*` files are good examples of this. I think we revisit this in follow up PRs as we flesh out the API we're exposing here in `internal/pkg/plugin`

mem replied:
Ok

Commit:
Change sylog.Fatalf to sylog.Errorf in plugin enable / disable

sylog.Fatalf will terminate the program. There are (exceptional) valid
use cases for that, but I think at this level, it's better to get into
the habit of using sylog.Errorf, as sylog.Fatalf will call os.Exit, and
os.Exit prevents defer'ed functions from running, which in this case
it probably does not make a difference, but the keyword here is
"probably" (as we don't know, in general, what the calling code is
doing).

Signed-off-by: Marcelo E. Magallon <marcelo@sylabs.io>

Commit:
Go back to sylog.Fatalf in plugin enable / disable

sylog.Fatalf calls os.Exit, meaning it immediately terminates the
program and does not allow defer'ed functions to run, etc. Since in
general we don't know what the calling code is doing, this is the wrong
thing to do, but it is the existing pattern. So for consistency's sake,
go back to that until we fix this properly everywhere.

Signed-off-by: Marcelo E. Magallon <marcelo@sylabs.io>
